java/arcs/core/data/proto/SchemaProtoDecoder.kt

```diff
@@ -0,0 +1,49 @@
+/*
+ * Copyright 2020 Google LLC.
+ *
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ *
+ * Code distributed by Google as part of this project is also subject to an additional IP rights
+ * grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+package arcs.core.data.proto
+
+import arcs.core.data.FieldName
+import arcs.core.data.FieldType
+import arcs.core.data.Schema
+import arcs.core.data.SchemaFields
+import arcs.core.data.SchemaName
+
+/**
+ * Returns the names in the [SchemaProto] as List<SchemaName>.
+ */
+fun SchemaProto.decodeNames(): List<SchemaName> = getNamesList().map { SchemaName(it) }
+
+/**
+ * Returns the fields in the [SchemaProto] as a Kotlin [SchemaFields] instance.
+ */
+fun SchemaProto.decodeFields(): SchemaFields {
+    val singletons = mutableMapOf<FieldName, FieldType>()
+    val collections = mutableMapOf<FieldName, FieldType>()
+    for ((name, type) in getFieldsMap()) {
+        when (type.getDataCase()) {
+            TypeProto.DataCase.PRIMITIVE -> singletons[name] = type.decodeAsFieldType()
+            // TODO: TypeProto.DataCase.COLLECTIONS
```

Comment:
+ TODO references

Reply:
Actually, `SchemaFields` has only two properties `singletons` and `collections`: https://github.com/PolymerLabs/arcs/blob/0f0f82a439cd00b2ea451f487bc3f96ff63635f1/java/arcs/core/data/SchemaFields.kt#L45-L48

What would a reference type map to? I believe it has to be a `singleton`?

Reply:
hmmm... schema can have a primitive field type which is a reference or a collection of references. I don't see how singleton relates to this.

Reply:
`Schema` class in Kotlin has a fields property, which is of type `SchemaFields` type shown avid. I am trying to determine how to translate the `fields` in `SchemaProto`:
```
message SchemaProto {
  repeated string names = 1;
  map<string, TypeProto> fields = 2;
}
```

Reply:
I just looked into SchemaFields, the reference over there is the EntityRef.

so if the type of the schema field in the proto is:
```
reference {
  referred_type {
    entity {
      schema {
       (...)
       hash: "abcd"
      }
    }
  }
}
```

it maps to `EntitiyRef("abcd")`

Reply:
Summarizing the offline discussion for the record. Collections of any type go into `collections`. Others go into `singletons`. We don't have support for tuples in Kotlin yet, and will be skipped for now.